kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -454,36 +451,36 @@ simplifyConjunctionByAssumption (toList -> andPredicates) =
 
     applyAssumptions
         ::  Predicate variable
-        ->  StateT (DoubleMap variable) Changed (Predicate variable)
+        ->  StateT (Assumptions variable) Changed (Predicate variable)
     applyAssumptions replaceIn = do
         assumptions <- State.get
         lift (applyAssumptionsWorker assumptions replaceIn)
 
     applyAssumptionsWorker
-        :: DoubleMap variable
+        :: Assumptions variable
         -> Predicate variable
         -> Changed (Predicate variable)
     applyAssumptionsWorker assumptions original
-      | Just result <- HashMap.lookup original (predMap assumptions) = Changed result
-
-      | HashMap.null (termLikeMap assumptions') &&
-        HashMap.null (predMap assumptions') = Unchanged original
-
-      | otherwise = (case replaceIn of
-          Predicate.CeilF ceil_ -> Predicate.CeilF <$> traverse
-            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) ceil_
-          Predicate.FloorF floor_ -> Predicate.FloorF <$> traverse
-            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) floor_
-          Predicate.EqualsF equals_ -> Predicate.EqualsF <$> traverse
-            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) equals_
-          Predicate.InF in_ -> Predicate.InF <$> traverse
-            (applyAssumptionsWorkerTerm (termLikeMap assumptions')) in_
-          _ -> traverse (applyAssumptionsWorker assumptions') replaceIn
-        )
-        & getChanged
-        -- The next line ensures that if the result is Unchanged, any allocation
-        -- performed while computing that result is collected.
-        & maybe (Unchanged original) (Changed . synthesize)
+      | Just result <- HashMap.lookup original (predicateMap assumptions) = Changed result
+
+      | HashMap.null (termLikeMap assumptions')
+      , HashMap.null (predicateMap assumptions') = Unchanged original
+
+      | otherwise =
+          case replaceIn of
+              Predicate.CeilF ceil_ -> Predicate.CeilF <$> traverse
+                (applyAssumptionsWorkerTerm (termLikeMap assumptions')) ceil_
```

Comment:
Let's add a local definition `applyAssumptionsWorkerTerm' = applyAssumptionsWorkerTerm (termLikeMap assumptions')` to avoid some of this duplication:

```suggestion
              Predicate.CeilF ceil_ ->
                  Predicate.CeilF <$> traverse applyAssumptionsWorkerTerm' ceil_                
```

Please check the indentation around the guards and case block here, too.